File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    int[] dr = {-1, 1, 0, 0};
    int[] dc = {0, 0, -1, 1};
    public int numIslands(char[][] grid) {

        int count = 0;
        

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    count++;

                    dfs(grid, i, j);
                   
                }
                
            }
        }

        return count;
    }

    private void dfs(char[][] grid, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] == '0') {
            return;
        }

        grid[i][j] = '0';

        for (int k = 0; k < 4; k++) {
            dfs(grid, i + dr[k], j + dc[k]);
        }
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        dr, dc = [1, -1, 0, 0], [0, 0, 1, -1] # up, down, right, left
        m, n = len(grid), len(grid[0])

        def dfs(row, col):
            grid[row][col] = "0"

            for k in range(4):
                nr, nc = row + dr[k], col + dc[k]

                if 0 <= nr < m and 0 <= nc < n and grid[nr][nc] == "1":
                    dfs(nr, nc)
        
        count = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    count += 1
                    dfs(i, j)

        return count